### ISSUE TITLE
Some Disambiguations are filtered out, why?

### ISSUE THREAD

**Issue by chile12** [dbpedia/extraction-framework#523]
I'm opening this in behalf of Martin Ostrovsky, reporting this issue via the mailing lists:

```
I’ve got the latest dump of disambiguation pages (2016-10) and I do a bzgrep on it for “Habs”, expecting to find the entries listed here:

https://en.wikipedia.org/wiki/Habs

but I only find these ones:

<http://dbpedia.org/resource/Habs> <http://dbpedia.org/ontology/wikiPageDisambiguates> <http://dbpedia.org/resource/Habitants> .
<http://dbpedia.org/resource/Habs> <http://dbpedia.org/ontology/wikiPageDisambiguates> <http://dbpedia.org/resource/Haberdashers'_Aske's_Boys'_School> .
<http://dbpedia.org/resource/Habs> <http://dbpedia.org/ontology/wikiPageDisambiguates> <http://dbpedia.org/resource/Historic_American_Buildings_Survey> .
<http://dbpedia.org/resource/Habs> <http://dbpedia.org/ontology/wikiPageDisambiguates> <http://dbpedia.org/resource/Trenton_Habs> .
```

In fact, there should be 8 disambiguations. The reason is this filter in the [DisambiguationExtractor](https://github.com/dbpedia/extraction-framework/blob/1f69f7ee042f05d6a0084f06b336cfe0eba4508f/core/src/main/scala/org/dbpedia/extraction/mappings/DisambiguationExtractor.scala#L40):
```
      // Extract only links that contain the page title or that spell out the acronym page title
      val disambigLinks = allLinks.filter { linkNode =>
        val cleanLink = linkNode.destination.decoded.toUpperCase(language.locale)
        cleanLink.contains(cleanPageTitle)|| isAcronym(cleanPageTitle, cleanLink)
      }
```
Can someone explain to me the reasoning behind this filter, or why we only allow variations of the original title or acronyms of it as Disambiguations?

**Comment by jcsahnwaldt** [comment]
First, a minor correction: It's correct that the current version of http://en.wikipedia.org/wiki/Habs has **eight** links to disambiguated pages, but the current version of http://dbpedia.org/resource/Habs (I guess it's the one from dump 2016-10) was derived from (prov:wasDerivedFrom) http://en.wikipedia.org/wiki/Habs?oldid=734591187, which only had **seven** links.

Now for the main problem. There are several things that make it hard for DBpedia to correctly recognize and parse Wikipedia disambiguation pages. One of the problems is that the links to the disambiguated pages are just normal links, and a disambiguation page may contain many other links. For example, see [Babs#In_entertainment](https://en.wikipedia.org/wiki/Babs#In_entertainment). There are eight links in the section, but only four of them lead to disambiguated pages. The other four lead to pages about films and shows in which a character named 'Babs' occurs. It would be wrong to generate DBpedia links like `dbr:Babs dbo:wikiPageDisambiguates dbr:Chicken_Run` etc. To avoid generating such wrong links, DBpedia decided years ago to use a heuristic to try to recognize the actual links to disambiguated pages and drop the others.

Basically, it's a trade-off between precision and recall. Maybe DBpedia is erring on the side of precision here. Maybe it would be possible to find a different heuristic. Maybe it would be better to drop this heuristic altogether and just accept the wrong links. Maybe there are only very few. Since I'm not involved with DBpedia anymore, I can't tell. It's even possible that Wikipedia disambiguation pages have evolved and are somewhat 'cleaner' now.

**Comment by chile12** [comment]
Thanks Christian for clarifying this for us. 

Alas, Wikipedia did not change anything regarding the wiki text structure of a disambiguation page and the target pages also rarely give an indication of its disambiguation status.

In the end it comes down to these questions:
1. What exactly is a disambiguation?
2. How strict do we want to be regarding the question 1. and the property dbo:wikiPageDisambiguates?
3. How many of these semantically dubious "disambiguations" would we create when removing this filter?

So in the example by Christian (Babs#In_entertainment): would we consider a disambiguation of the name 'Babs' to a movie with a character called Babs really such a 'wrong' link?

The last question is easy to answer. The others need some discussion.

**Comment by jimkont** [comment]
@jcsahnwaldt nailed it in detail. I would be in favor of  keeping the configuration as it is and favor precision instead of recall.

With a quick check at https://en.wikipedia.org/wiki/Apple_(disambiguation) we would get pages like tomato which does not seem like a disambiguation of apple

**Comment by mgns** [comment]
Maybe OT, but related: It should also be clarified how disambiguations are interpreted in a Fusion dataset. Word ambiguities are strongly language dependent, this should potentially be solved on a literal / language level.

**Comment by chile12** [comment]
Ok, to summarize:

- we keep the filter
- I will look into extending it if possible

@mgns point well taken, yet I fear this is a task we would have to outsource to the chapters.
In general, I would argue for adding a hook for postprocessing in the extractor interface:

(quads: Seq[Quad], graphIri: String, subjectIri: String) => Seq[Quad]

or similar...

**Comment by chile12** [comment]
Ok so I have widened the filter as follows:

```        
        val cleanLink = linkNode.destination.decoded.toUpperCase(language.locale)
        val lineText = linkNode.root.getOriginWikiText(linkNode.line).toUpperCase(language.locale).trim
        val titleSplits = cleanPageTitle.split(" ")

        // link text contains at least half of all word in the title
        val condition1a = titleSplits.count(s => cleanLink.contains(s)) >= titleSplits.size.toDouble / 2d
        // the link appears at the start of the line and the line contains more than half of the title words
        val condition1b = lineText.matches("^\\s*\\*\\s*\\[\\[" + cleanLink + "\\]\\].*") && titleSplits.count(s => lineText.contains(s)) > titleSplits.size.toDouble / 2d
        // acronym appears in the whole line
        val condition1c = isAcronym(cleanPageTitle, lineText)
        // disallow redirects to page sections
        val condition2 = ! cleanLink.contains("#")

        condition2 && (condition1a || condition1b || condition1c)
```

Reasoning:
condition1a: Important if the disamb title has more than one word. For example https://en.wikipedia.org/wiki/Service_worker will now produce all 3 redirects.

condition1b: extends the search for the title words to the whole line where the link appears, but only if the link is at the start of the line. Often you will find that links to "unrelated" or explanatory subjects appear further down the line, direct links are often at the start of the line (e.g. [Set](https://en.wikipedia.org/w/index.php?title=Set#Sport_and_games).

condition1c: extended the search for acronyms to the whole line (useful e.g. for [Single Embryo Transfer](https://en.wikipedia.org/w/index.php?title=Set#Acronyms)). 

condition2: I will not redirect to targets with fragments.

**Comment by chile12** [comment]
Closing
